File: Sources/XDecodeCore/DecodeCoordinator.swift
import Darwin
import Foundation

public typealias DecoderResolver = @Sendable (DecodeRequest) async throws -> any LogDecoder

public actor DecodeCoordinator {
    private let fileManager: FileManager
    private let decoderResolver: DecoderResolver

    public init(
        fileManager: FileManager = .default,
        decoderResolver: @escaping DecoderResolver
    ) {
        self.fileManager = fileManager
        self.decoderResolver = decoderResolver
    }

    public func decode(_ request: DecodeRequest) async -> DecodeResult {
        let sourceURL = request.sourceURL.standardizedFileURL
        let directory = sourceURL.deletingLastPathComponent()
        let temporaryURL = directory.appendingPathComponent(".xdecode-\(UUID().uuidString).tmp")
        var outputURL: URL?

        do {
            guard fileManager.fileExists(atPath: sourceURL.path) else {
                throw DecodeError.fileOperation("源文件不存在")
            }

            let input = try Data(contentsOf: sourceURL, options: .mappedIfSafe)
            let decoder = try await decoderResolver(request)
            let decoded = try decoder.decode(input, sourceURL: sourceURL)

            guard decoded.isComplete else {
                return DecodeResult(
                    request: request,
                    state: .failed,
                    outputURL: nil,
                    message: "解密失败：\(decoded.diagnostic ?? "日志未完整解密")；未生成 .log，源文件已保留",
                    sourceDeleted: false
                )
            }
            guard !decoded.data.isEmpty else { throw DecodeError.emptyOutput }

            try decoded.data.write(to: temporaryURL)
            let handle = try FileHandle(forWritingTo: temporaryURL)
            try handle.synchronize()
            try handle.close()

            outputURL = try moveToUniqueOutput(temporaryURL: temporaryURL, sourceURL: sourceURL)

            do {
                try fileManager.removeItem(at: sourceURL)
                let message = decoded.diagnostic.map { "\($0)；源文件已永久删除" }
                    ?? "解密完成，源文件已永久删除"
                return DecodeResult(
                    request: request,
                    state: .completed,
                    outputURL: outputURL,
                    message: message,
                    sourceDeleted: true
                )
            } catch {
                let message = decoded.diagnostic.map {
                    "\($0)；源文件删除失败：\(error.localizedDescription)"
                } ?? "解密完成，但源文件删除失败：\(error.localizedDescription)"
                return DecodeResult(
                    request: request,
                    state: .completedWithWarning,
                    outputURL: outputURL,
                    message: message,
                    sourceDeleted: false
                )
            }
        } catch {
            try? fileManager.removeItem(at: temporaryURL)
            return DecodeResult(
                request: request,
                state: .failed,
                outputURL: outputURL,
                message: error.localizedDescription,
                sourceDeleted: false
            )
        }
    }

    public static func outputURL(for sourceURL: URL, index: Int = 0) -> URL {
        let directory = sourceURL.deletingLastPathComponent()
        let baseName = sourceURL.deletingPathExtension().lastPathComponent
        let suffix = index == 0 ? "" : "-\(index)"
        return directory.appendingPathComponent("\(baseName)\(suffix).log", isDirectory: false)
    }

    private func moveToUniqueOutput(temporaryURL: URL, sourceURL: URL) throws -> URL {
        var index = 0
        while true {
            let candidate = Self.outputURL(for: sourceURL, index: index)
            let result = temporaryURL.path.withCString { sourcePath in
                candidate.path.withCString { destinationPath in
                    renamex_np(sourcePath, destinationPath, UInt32(RENAME_EXCL))
                }
            }
            if result == 0 {
                return candidate
            }

            let renameError = errno
            if renameError == EEXIST {
                index += 1
            } else {
                let message = String(cString: strerror(renameError))
                throw DecodeError.fileOperation(message)
            }
        }
    }
}

public enum StandardDecoderResolver {
    public static func make(
        xlogCredentials: @escaping @Sendable (URL) async -> [XlogCredentials],
        loganCredentials: @escaping @Sendable (URL) async throws -> [LoganCredentials]
    ) -> DecoderResolver {
        { request in
            switch request.format {
            case .xlog:
                return XlogDecoder(credentials: await xlogCredentials(request.sourceURL))
            case .mx:
                return MXDecoder()
            case .logan:
                return LoganDecoder(credentials: try await loganCredentials(request.sourceURL))
            case .zip:
                throw DecodeError.unsupportedFormat("ZIP 容器需要通过批量解密流程处理")
            }
        }
    }
}

File: Windows/src/XDecode.Core/DecodeCoordinator.cs
namespace XDecode.Core;

public sealed class DecodeCoordinator(DecoderResolver decoderResolver)
{
    public async Task<DecodeResult> DecodeAsync(
        DecodeRequest request,
        CancellationToken cancellationToken = default)
    {
        var sourcePath = Path.GetFullPath(request.SourcePath);
        var directory = Path.GetDirectoryName(sourcePath)
            ?? throw DecodeException.FileOperation("无法确定源文件目录");
        var temporaryPath = Path.Combine(directory, $".xdecode-{Guid.NewGuid():N}.tmp");
        string? outputPath = null;
        try
        {
            if (!File.Exists(sourcePath)) throw DecodeException.FileOperation("源文件不存在");
            DecodeLimits.ValidateInputFile(sourcePath, "源文件");
            var input = await File.ReadAllBytesAsync(sourcePath, cancellationToken).ConfigureAwait(false);
            var decoder = await decoderResolver(request, cancellationToken).ConfigureAwait(false);
            var decoded = decoder.Decode(input, sourcePath);
            if (!decoded.IsComplete)
            {
                return new DecodeResult(
                    request, DecodeState.Failed, null,
                    $"解密失败：{decoded.Diagnostic ?? "日志未完整解密"}；未生成 .log，源文件已保留",
                    false);
            }
            if (decoded.Data.Length == 0) throw DecodeException.EmptyOutput();
            DecodeLimits.ValidateOutputSize(0, decoded.Data.Length);

            await using (var stream = new FileStream(
                temporaryPath, FileMode.CreateNew, FileAccess.Write, FileShare.None,
                64 * 1024, FileOptions.Asynchronous | FileOptions.WriteThrough))
            {
                WindowsFilePublication.MarkHidden(temporaryPath);
                await stream.WriteAsync(decoded.Data, cancellationToken).ConfigureAwait(false);
                await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
                stream.Flush(flushToDisk: true);
            }
            outputPath = MoveToUniqueOutput(temporaryPath, sourcePath);

            try
            {
                File.Delete(sourcePath);
                return new DecodeResult(
                    request, DecodeState.Completed, outputPath,
                    decoded.Diagnostic is null
                        ? "解密完成，源文件已永久删除"
                        : $"{decoded.Diagnostic}；源文件已永久删除",
                    true);
            }
            catch (Exception exception)
            {
                return new DecodeResult(
                    request, DecodeState.CompletedWithWarning, outputPath,
                    decoded.Diagnostic is null
                        ? $"解密完成，但源文件删除失败：{exception.Message}"
                        : $"{decoded.Diagnostic}；源文件删除失败：{exception.Message}",
                    false);
            }
        }
        catch (Exception exception)
        {
            TryDelete(temporaryPath);
            return new DecodeResult(
                request, DecodeState.Failed, outputPath,
                exception.Message, false);
        }
    }

    public static string OutputPath(string sourcePath, int index = 0)
    {
        var directory = Path.GetDirectoryName(sourcePath) ?? "";
        var baseName = Path.GetFileNameWithoutExtension(sourcePath);
        var suffix = index == 0 ? "" : $"-{index}";
        return Path.Combine(directory, $"{baseName}{suffix}.log");
    }

    private static string MoveToUniqueOutput(string temporaryPath, string sourcePath)
    {
        for (var index = 0; ; index++)
        {
            var candidate = OutputPath(sourcePath, index);
            if (WindowsFilePublication.TryMoveExclusive(temporaryPath, candidate)) return candidate;
        }
    }

    private static void TryDelete(string path)
    {
        try { if (File.Exists(path)) File.Delete(path); }
        catch { }
    }
}

public static class StandardDecoderResolver
{
    public static DecoderResolver Create(
        Func<string, CancellationToken, ValueTask<IReadOnlyList<XlogCredentials>>> xlogCredentials,
        Func<string, CancellationToken, ValueTask<IReadOnlyList<LoganCredentials>>> loganCredentials) =>
        async (request, cancellationToken) => request.Format switch
        {
            LogFormat.Xlog => new XlogDecoder(await xlogCredentials(request.SourcePath, cancellationToken)),
            LogFormat.Mx => new MxDecoder(),
            LogFormat.Logan => new LoganDecoder(await loganCredentials(request.SourcePath, cancellationToken)),
            _ => throw DecodeException.UnsupportedFormat("ZIP 容器需要通过批量解密流程处理")
        };
}
